### DISCUSSION TITLE
UUID is one of $CURRENT_USER.schools as item permission rule throws an Postgres error

### DISCUSSION THREAD

**Discussion by viu-x** [directus/directus#17585] [Archive]
### Checklist

- [X] I'm using the [latest version of Directus](https://github.com/directus/directus/releases)
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes the problem.
- [X] I've completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).

### Describe the Bug

I've added a custom field to `directus_users` named `schools`. It's an M2M relation to my collection `school` with an ID of type UUID. And I want to restrict access to the `school` items which are assigned to a user. Therefore I created a role where I set custom read access to that collection with a rule in item permissions:

![Bildschirm­foto 2023-02-19 um 21 55 46](https://user-images.githubusercontent.com/103430706/219974768-765c216c-1613-4a15-9050-5b32b3dcc253.png)

When I access the collection with my test user having that role and two schools assigned the console show me this error

```
ERROR: select count(*) as "count" from "school" where (("school"."id" in ($1, $2))) limit $3 - invalid input syntax for type uuid: "3"
    err: {
      "type": "DatabaseError",
      "message": "select count(*) as \"count\" from \"school\" where ((\"school\".\"id\" in ($1, $2))) limit $3 - invalid input syntax for type uuid: \"3\"",
      "stack":
          error: select count(*) as "count" from "school" where (("school"."id" in ($1, $2))) limit $3 - invalid input syntax for type uuid: "3"
              at Parser.parseErrorMessage (/Users/tpp/_projects/example-project/node_modules/pg-protocol/dist/parser.js:287:98)
              at Parser.handlePacket (/Users/tpp/_projects/example-project/node_modules/pg-protocol/dist/parser.js:126:29)
              at Parser.parse (/Users/tpp/_projects/example-project/node_modules/pg-protocol/dist/parser.js:39:38)
              at Socket.<anonymous> (/Users/tpp/_projects/example-project/node_modules/pg-protocol/dist/index.js:11:42)
              at Socket.emit (node:events:512:28)
              at addChunk (node:internal/streams/readable:324:12)
              at readableAddChunk (node:internal/streams/readable:297:9)
              at Readable.push (node:internal/streams/readable:234:10)
              at TCP.onStreamRead (node:internal/stream_base_commons:190:23)
      "length": 96,
      "name": "error",
      "severity": "ERROR",
      "code": "22P02",
      "file": "uuid.c",
      "line": "134",
      "routine": "string_to_uuid"
    }
```

The query seems to be correct. When I copy paste it and insert the corresponding UUIDs it returns the correct count of 2.

```sql
SELECT
	count(*) AS "count"
FROM
	"school"
WHERE (("school"."id" in('14f64659-b640-4241-a04a-a15ff81e01bb', '9ed9ebac-3f66-489d-bfb4-c8b672a0cb91')))
```

### To Reproduce

1. Create a collection with UUID primary key
2. Create some entries
3. Add a M2M field to directus_users which has a relation to that collection
4. Create a role and set custom access to that collection, with an item permissions rule `ID is one of $CURRENT_USER.your_field`
5. Create a user and assign that new role and some entries
6. Log in with that user and access the collection

You should an error message like above in your server console and a view on the collection page like the following

![Bildschirm­foto 2023-02-19 um 22 14 20](https://user-images.githubusercontent.com/103430706/219975602-6852c991-142c-4c5f-b161-edc183c51c44.png)

### Hosting Strategy

Self-Hosted (Docker Image)

**Comment by paescuj** [comment]
This sounds more like a configuration problem to me. Happy to transform it back to an issue if it turns out that this is not the case!

**Reply by paescuj** [reply]
So here's what I meant by configuration problem: Since you're using a M2M field in `directus_users` the IDs of the junction table is stored in this field and not the actual IDs of the `school` collection, which means in this case your permission rule is not quite suitable.
You might want to use a O2M field in `directus_users` instead, which would work with the exact same permission rule.

However, I can see how this does not necessarily meet expectations. So I'm happy to discuss possible measures / solutions here.

**Reply by viu-x** [reply]
I tried O2M at first but it is not possible to create. I can't select any field from the related collection. Doesn't matter which collection I try. 

![Bildschirm­foto 2023-02-21 um 15 41 14](https://user-images.githubusercontent.com/103430706/220375969-557304c7-7acc-49d2-89e5-48ef6d022647.png)

I even read the docs for O2M again and created an alias field inside my `school` collection. But that field is also not selectable.

But then again is O2M not the correct model in the end. Because every `user` can access multiple `schools` and one `school` can be managed by multiple `users`.

**Reply by paescuj** [reply]
You would have to create an own field (e.g. `users`) for the foreign key instead of selecting an existing one.

> But then again is O2M not the correct model in the end. Because every user can access multiple schools and one school can be managed by multiple users.

Okay, I see 👍

**Reply by viu-x** [reply]
So is there any possibility to achieve our scenario? To filter (multiple) collections (the ones having a relation to `school`) based the selected `school`(s) in the `directus_user` item?

That kinda makes or breaks the usage of directus for us in the long term.

**Reply by paescuj** [reply]
Let's figure out potential solutions with their advantages and disadvantages! 👍 
A few ideas from my side:

Further based on this set-up: `directus_user.schools <- M2M -> schools`

- Direct interpretation of the relation field to get to the relation: `$CURRENT_USER.schools`
  May be in line with expectation, but not possible to select fields from the junction table.
- Allow getting field in junction table for each entry in relation field: `$CURRENT_USER.schools.school_id`
  Still pretty simple but less intuitive, requires additional filter logic.
- Same as above but with JSONPath like syntax: `$CURRENT_USER.schools[*].school_id`
  Might be a bit more intuitive, but means even more complexity.

Of course, completely different solutions are also conceivable...

**Reply by toomuchdesign** [reply]
So let's consider the hardest scenario: many-to-many relationship from both sides:

- users has many school and vice versa
- or: user and content entity to be filtered have both many categories

If we considered extending the current filtering language we would need to:
- get a field in junction table for each entry in relation field
- extend [filter operators](https://docs.directus.io/reference/filter-rules.html#filter-operators) to implement a new operator which compares 2 set of values (or a set of values vs a value) to find at least one common match (something like `hasOne`)

In the permission filter:
<img width="504" alt="Screenshot 2023-02-24 at 15 07 06" src="https://user-images.githubusercontent.com/4573549/221198660-5d2900bb-e5e9-4dc1-a8f4-8e2566452208.png">
...we might be able to write something like:

```
ID is one of $CURRENT_USER.schools.ID
```
or:
```
category -> ID in one of $CURRENT_USER.categories.ID
```
Directus underneath should be able to interpret the `.ID` as a collection. Makes any sense?

**Comment by toomuchdesign** [comment]
Sorry for crashing in, I believe I'm also trying to achieve something similar: setting permissions based on **current user** and **content entity** relational fields where:

- **Current user** has a **many-to-many** relationship with a **category** entity
- **Content entity** has a **one-to-many** relationship with the same **category** entity

In other words I want users to be able to see the content entities associated to the same category they are associated to. User can have multiple categories.

Is there any way to set such permissions? Even via code, if necessary.

Thank you for the outstanding work here!
